Disconnect clients (without subscriptions) with same client id in a clustered setup

dergraf opened the pull request.

dergraf committed:
New subscriber DB format, streamlining the work with subscriber data

dergraf committed:
New subscriber DB format, streamlining the work with subscriber data

dergraf committed:
Added incremental data migration

larshesel commented on src/vmq_subscriber.erl:
```diff
@@ -0,0 +1,226 @@
+%% Copyright 2014 Erlio GmbH Basel Switzerland (http://erl.io)
```
change to 2016?

dergraf replied:
done

larshesel commented on src/vmq_subscriber_db.erl:
```diff
@@ -0,0 +1,73 @@
+%% Copyright 2014 Erlio GmbH Basel Switzerland (http://erl.io)
```
change copyright year here as well

dergraf replied:
done

larshesel commented on src/vmq_reg.erl:
```diff
@@ -332,7 +328,8 @@ deliver_retained({MP, _} = SubscriberId, Topic, QoS) ->
       end, ok, MP, Topic).
 
 subscriptions_for_subscriber_id(SubscriberId) ->
-    plumtree_metadata:get(?SUBSCRIBER_DB, SubscriberId, [{default, []}]).
```
In the `vmq_subscriber_db:read` call we no longer pass the `[{default, []}]` proplist to `plumtree_metadata:get/X`. I suppose that's fine (it doesn't change anything anyway)?

dergraf replied:
the option `[{default, []}]` is passed to tell plumtree to return the `[]`  instead of undefined if no value can be found for the given key. We handle this manually in `vmq_subscriber_db`.

larshesel commented on src/vmq_reg.erl:
```diff
@@ -636,88 +604,54 @@ fold_sessions(FoldFun, Acc) ->
+    case vmq_subscriber:add(OldSubs, [{T, QoS}||{T, QoS} <- Topics, is_integer(QoS)]) of
+        {NewSubs, true} ->
+            vmq_subscriber_db:store(SubscriberId, NewSubs);
+        _ ->
```
This change I wasn't able to follow - it's not clear to me that the removed fold above and is equivalent to the `vmq_subscriber:add/2`. I guess I need to spend more time understanding the new code there.

dergraf replied:
`vmq_subscriber:add/2` does essentially the same, just in a different more elegant way :)

larshesel reviewed (commented):
LGTM, just a couple of minor points and one things I didn't understand yet,...

dergraf committed:
change copyright year

larshesel commented:
I can confirm this fixes erlio/vernemq#201 for me but not erlio/vernemq#202. Regardless of the clean session flag, my client still receives any offline messages for my client id.

larshesel commented:
Shouldn't we be throwing out all data in vmq_reg:register_subscriber/x when cs=true instead of starting queue migration? Or should we do the queue migration and then when the queue is ready throw the data msgs away?

dergraf commented:
Ou, you're right. I was wrong thinking it belongs to the same issue.

dergraf committed:
Store clean session inside subscriber entry

This is required that the asynchronous migration mechanism can
distinguish between cleaning up an existing remote client or forcing the
queue migration.

dergraf committed:
Store clean session inside subscriber entry

This is required that the asynchronous migration mechanism can
distinguish between cleaning up an existing remote client or forcing the
queue migration.

dergraf committed:
Minor robustness improvements.

larshesel commented:
Hey @dergraf I'm reviewing the code, but tested it out first. I still receive offline messages even if I connect with clean session = true. Messages will no longer be stored after connecting with clean session = true, though. Here's a shell session showing what I mean - maybe I'm misunderstanding something:

``` shell
# Create a 'subber2' cs=false session
$ mosquitto_sub -c  -i subber2 -t "topic" -q 1
^C
$ mosquitto_pub -i myclientid -t "topic" -m "hello xcv" -q 1
# Retrieve the offline message
$ mosquitto_sub -c  -i subber2 -t "topic" -q 1
hello xcv
^C

# Let's publish a message again
$ mosquitto_pub -i myclientid -t "topic" -m "hello xcv" -q 1
# connect with cs=true: We still receive the offline message! So state was not discarded upon connection
$ mosquitto_sub -i subber2 -t "topic" -q 1
hello xcv
^C
$ mosquitto_pub -i myclientid -t "topic" -m "hello xcv" -q 1
# Second connection with cs=true - no messages was stored
$ mosquitto_sub -i subber2 -t "topic" -q 1
```

dergraf commented:
Yes, this is wrong. I've thought we have covered this case in `test/vmq_clean_session_SUITE.erl`, but apparently not.

dergraf committed:
Cleanup offline Queue in case a reconnecting client uses clean-session=true

dergraf committed:
Cleanup offline Queue in case a reconnecting client uses clean-session=true

dergraf committed:
use new vmq_queue_sup_sup module

larshesel commented on src/vmq_reg.erl:
```diff
@@ -395,8 +399,10 @@ fix_dead_queue(SubscriberId, Subs, {DeadNodes, [Target|Targets], N}) ->
-              vmq_subscriber:change_node(Acc, DeadNode, Target)
+      fun(DeadNode, AccSubs) ->
+              %% if the target node already has a 'new' session for
+              %% this subscriber, and old
```
comment incomplete?

larshesel commented on src/vmq_queue.erl:
```diff
@@ -516,6 +541,15 @@ handle_session_down(SessionPid, StateName,
+            cleanup_queue(SId, Q),
+            _ = vmq_metrics:incr_queue_unhandled(queue:len(Q)),
+            gen_fsm:reply(From, ok),
+            _ = vmq_plugin:all(on_client_gone, [SId]),
```
why is it `on_client_gone` in this case?

dergraf replied:
`on_client_gone` means the client is gone and all it's data with it.. as opposed to `on_client_offline` which is used in case we still have some client data.

larshesel commented on src/vmq_reg.erl:
```diff
@@ -634,18 +628,18 @@ del_subscriptions(Topics, SubscriberId) ->
         undefined ->
             %% Store empty Subscriber Data
-            Subs = vmq_subscriber:new(),
+            Subs = vmq_subscriber:new(false),
```
Maybe it would be nicer to match the clean session flag from the map and pass it on to the `vmq_subscriber:new/1` here and above as well? I was looking at the code and had to think twice why it was true/false in the two cases.

dergraf committed:
fix check_format eunit test

larshesel commented:
Shouldn't we be throwing out all data in vmq_reg:register_subscriber/x when cs=true instead of starting queue migration? Or should we do the queue migration and then when the queue is ready throw the data msgs away?

dergraf commented:
@larshesel this is done here https://github.com/erlio/vmq_server/pull/100/files#diff-5b39f0e2bd578539b1ddfa11520f3036R154

larshesel commented:
hey @dergraf sorry about the last comment - it was one I thought I'd submitted with the PR review last week - no idea why it first showed up 4 hours ago...

dergraf committed:
Added comments and use proper variable naming

larshesel commented:
LGTM
